context_legacy.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2022-2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !go1.21
+
+package otelconnect
+
+import (
+	"context"
+	"sync"
+)
+
+// afterFunc is a simple imitation of context.AfterFunc from Go 1.21.
+// It is not as efficient as the real implementation, but it is sufficient
+// for our purposes.
+func afterFunc(ctx context.Context, f func()) (stop func() bool) {
+	ctx, cancel := context.WithCancel(ctx)
+	var once sync.Once
```

Comment:
This seems a little overkill, especially since it means that if the goroutine "wins", the call to `stop` will block for `f` to complete.

The actual implementation of `context.AfterFunc` explicitly states that `stop` does **not** wait for `f` to complete:

> Calling the returned stop function stops the association of ctx with f. It returns true if the call stopped f from being run. If stop returns false, either the context is done and f has been started in its own goroutine; or f was already stopped. The stop function does not wait for f to complete before returning. If the caller needs to know whether f is completed, it must coordinate with f explicitly.

So I think an atomic bool is a more appropriate guard here.

Reply:
Switched to atomic bool.